server/src/main/java/org/elasticsearch/cluster/routing/ExpectedShardSizeEstimator.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.snapshots.SnapshotShardSizeInfo;
+
+import java.util.Set;
+
+public class ExpectedShardSizeEstimator {
+
+    public static long getExpectedShardSize(ShardRouting shardRouting, long defaultSize, RoutingAllocation allocation) {
+        return getExpectedShardSize(
+            shardRouting,
+            defaultSize,
+            allocation.clusterInfo(),
+            allocation.snapshotShardSizeInfo(),
+            allocation.metadata(),
+            allocation.routingTable()
+        );
+    }
+
+    /**
+     * Returns the expected shard size for the given shard or the default value provided if not enough information are available
+     * to estimate the shards size.
+     */
+    public static long getExpectedShardSize(
+        ShardRouting shard,
+        long defaultValue,
+        ClusterInfo clusterInfo,
+        SnapshotShardSizeInfo snapshotShardSizeInfo,
+        Metadata metadata,
+        RoutingTable routingTable
+    ) {
+        final IndexMetadata indexMetadata = metadata.getIndexSafe(shard.index());
+        if (indexMetadata.getResizeSourceIndex() != null
+            && shard.active() == false
+            && shard.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
+            return getExpectedSizeOfClonedShard(shard, defaultValue, indexMetadata, clusterInfo, metadata, routingTable);
+        } else if (shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+            return snapshotShardSizeInfo.getShardSize(shard, defaultValue);
+        } else {
+            return clusterInfo.getShardSize(shard, defaultValue);
+        }
+    }
+
+    private static long getExpectedSizeOfClonedShard(
```

Comment:
nit: this is for all resize operations (clone, shrink and split)

```suggestion
    private static long getExpectedSizeOfResizedShard(
```